{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook includes generation of the following protein features:\n",
    "- Basic sequence features (25)\n",
    "    - Sequence length (1)\n",
    "    - Molecular weight (1)\n",
    "    - Amino acid count (20)\n",
    "    - hydrophobic amino acid count (1)\n",
    "    - hydrophobic amino acid count on surface (1)\n",
    "    - polar amino acid count (1)\n",
    "- Global features derived from NetSurfP-2.0 predictions (7)\n",
    "    - proportion of helix structure\n",
    "    - proportion of turn structure\n",
    "    - proportion of sheet structure\n",
    "    - proportion of disordered regions\n",
    "    - total accessible surface area\n",
    "    - total hydrophobic surface area\n",
    "    - relative hydrophobic surface area\n",
    "    - Exposed amino acid count (20)\n",
    "- Solubility-weighted index (2)\n",
    "    - SWI\n",
    "    - Probability of Solubility\n",
    "- Aggregation propensity (1)\n",
    "- Simple protein analysis (BioPython module) (6)\n",
    "    - Aromaticity\n",
    "    - Instability index\n",
    "    - Gravy\n",
    "    - Isoelectric point\n",
    "    - Charge at pH 5\n",
    "    - Charge at pH 7\n",
    "- HSP annotation (1)\n",
    "- PTM annotations (15)\n",
    "    - PTM (UniProt)\n",
    "    - Acetylation (combined)\n",
    "    - Glycosylation (combined)\n",
    "    - Methylation (combined)\n",
    "    - Myristoylation (combined)\n",
    "    - Nitrosylation (combined)       \n",
    "    - Palmitoylation (combined)\n",
    "    - Phosphorylation (combined)\n",
    "    - SUMOylation (combined)\n",
    "    - Ubiquitination (combined)\n",
    "    - Citrullination (UniProt)\n",
    "    - GPI-anchor (UniProt)\n",
    "    - Lipoprotein (UniProt)\n",
    "    - Nitration (UniProt)\n",
    "    - Prenylation (UniProt)\n",
    "- PTM predictions (10)\n",
    "    - PTM_count\t\n",
    "    - Phosphorylation\t\n",
    "    - Glycosylation\n",
    "    - Ubiquitination\n",
    "    - SUMOylation\n",
    "    - Acetylation\n",
    "    - Palmitoylation\n",
    "    - Hydroxylation\n",
    "    - Methylation\n",
    "    - Pyrrolidone_carboxylic_acid\n",
    "- PROSITE domains (5)\n",
    "    - Coiled coil domain\n",
    "    - WW domain\n",
    "    - RAS profile\n",
    "    - EGF\n",
    "    - RRM\n",
    "- Transmembrane annotations (1)\n",
    "- Transmembrane predictions (1)\n",
    "    \n",
    "This notebook also includes:\n",
    "- Data normalization (amino acid count values to proportions)\n",
    "- Log transformation of features\n",
    "- Removal of non-MS detectable proteins from the dataset\n",
    "       \n",
    "Output dataset:\n",
    "- features_human_proteome.csv\n",
    "- features_human_proteome_no_filtering.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from os.path import isfile, join"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "Data_path = os.path.dirname(os.getcwd()) + '/Data'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NetSurfP-2.0 predictions\n",
    "\n",
    "Dataset already contains:\n",
    "- Sequence length\n",
    "- Molecular weight\n",
    "- Amino acid count\n",
    "- NetSurfP-2.0 predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import NSP2 predictions for human genome\n",
    "nsp = Data_path + '/Raw/NSP2_complete.tab'\n",
    "df_features = pd.read_csv(nsp, sep='\\t') # 20384 proteins, does not include TITIN (NSP2 cannot run such a big protein)\n",
    "df_features.drop(columns=['PDB', 'q3_H', 'q3_E', 'q3_C', 'on_surface'], inplace=True) # drop ununsed columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>length</th>\n",
       "      <th>hydr_count</th>\n",
       "      <th>polar_count</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>helix</th>\n",
       "      <th>turn</th>\n",
       "      <th>sheet</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>...</th>\n",
       "      <th>S</th>\n",
       "      <th>T</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>fasta_sequence</th>\n",
       "      <th>thsa_netsurfp2</th>\n",
       "      <th>tasa_netsurfp2</th>\n",
       "      <th>rhsa_netsurfp2</th>\n",
       "      <th>disorder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8N7X0</td>\n",
       "      <td>1667</td>\n",
       "      <td>627</td>\n",
       "      <td>719</td>\n",
       "      <td>219717.24</td>\n",
       "      <td>0.225555</td>\n",
       "      <td>0.604079</td>\n",
       "      <td>0.170366</td>\n",
       "      <td>95</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>154</td>\n",
       "      <td>104</td>\n",
       "      <td>95</td>\n",
       "      <td>28</td>\n",
       "      <td>49</td>\n",
       "      <td>MASKQTKKKEVHRINSAHGSDKSKDFYPFGSNVQSGSTEQKKGKFP...</td>\n",
       "      <td>25102.465221</td>\n",
       "      <td>108276.162948</td>\n",
       "      <td>0.231837</td>\n",
       "      <td>0.335933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q5T1N1</td>\n",
       "      <td>836</td>\n",
       "      <td>243</td>\n",
       "      <td>401</td>\n",
       "      <td>107902.81</td>\n",
       "      <td>0.183014</td>\n",
       "      <td>0.777512</td>\n",
       "      <td>0.039474</td>\n",
       "      <td>44</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>91</td>\n",
       "      <td>55</td>\n",
       "      <td>24</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>MDEADFSEHTTYKQEDLPYDGDLSQIKIGNDYSFTSKKDGLEVLNQ...</td>\n",
       "      <td>14688.543593</td>\n",
       "      <td>67203.514041</td>\n",
       "      <td>0.218568</td>\n",
       "      <td>0.533493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q92667</td>\n",
       "      <td>903</td>\n",
       "      <td>340</td>\n",
       "      <td>335</td>\n",
       "      <td>113584.31</td>\n",
       "      <td>0.079734</td>\n",
       "      <td>0.805094</td>\n",
       "      <td>0.115172</td>\n",
       "      <td>72</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>94</td>\n",
       "      <td>49</td>\n",
       "      <td>72</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>MAIQFRSLFPLALPGMLALLGWWWFFSRKKGHVSSHDEQQVEAGAV...</td>\n",
       "      <td>19982.520280</td>\n",
       "      <td>68025.600814</td>\n",
       "      <td>0.293750</td>\n",
       "      <td>0.644518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q5VUY0</td>\n",
       "      <td>407</td>\n",
       "      <td>208</td>\n",
       "      <td>127</td>\n",
       "      <td>53468.44</td>\n",
       "      <td>0.491400</td>\n",
       "      <td>0.380835</td>\n",
       "      <td>0.127764</td>\n",
       "      <td>24</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>14</td>\n",
       "      <td>39</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>MWDLALIFLAAACVFSLGVTLWVICSHFFTVHIPAAVGHPVKLRVL...</td>\n",
       "      <td>6416.179985</td>\n",
       "      <td>18254.852216</td>\n",
       "      <td>0.351478</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P62736</td>\n",
       "      <td>377</td>\n",
       "      <td>161</td>\n",
       "      <td>143</td>\n",
       "      <td>48781.11</td>\n",
       "      <td>0.445623</td>\n",
       "      <td>0.347480</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>29</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>26</td>\n",
       "      <td>24</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>MCEEEDSTALVCDNGSGLCKAGFAGDDAPRAVFPSIVGRPRHQGVM...</td>\n",
       "      <td>4444.627391</td>\n",
       "      <td>17613.271161</td>\n",
       "      <td>0.252345</td>\n",
       "      <td>0.037135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20379</th>\n",
       "      <td>A0PK05</td>\n",
       "      <td>275</td>\n",
       "      <td>126</td>\n",
       "      <td>89</td>\n",
       "      <td>34825.86</td>\n",
       "      <td>0.316364</td>\n",
       "      <td>0.683636</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>24</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>MQLQVFWTGLEYTCRLLGITTAAVLIGVGTETFLQGQFKSLAFYLL...</td>\n",
       "      <td>10264.757559</td>\n",
       "      <td>23347.907188</td>\n",
       "      <td>0.439644</td>\n",
       "      <td>0.483636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20380</th>\n",
       "      <td>Q9HCN2</td>\n",
       "      <td>124</td>\n",
       "      <td>46</td>\n",
       "      <td>35</td>\n",
       "      <td>15149.80</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MGSSSEASFRSAQASCSGARRQGLGRGDQNLSVMPPNGRAQTHTPG...</td>\n",
       "      <td>2739.538957</td>\n",
       "      <td>8771.609425</td>\n",
       "      <td>0.312319</td>\n",
       "      <td>0.048387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20381</th>\n",
       "      <td>A0A0A0MS03</td>\n",
       "      <td>114</td>\n",
       "      <td>47</td>\n",
       "      <td>36</td>\n",
       "      <td>14631.45</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.535088</td>\n",
       "      <td>0.438596</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>MGPGLLCWELLYLLGAGPVEAGVTQSPTHLIKTRGQQVTLRCSPIS...</td>\n",
       "      <td>2996.623152</td>\n",
       "      <td>7889.445581</td>\n",
       "      <td>0.379827</td>\n",
       "      <td>0.184211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20382</th>\n",
       "      <td>A0A0A6YYK4</td>\n",
       "      <td>115</td>\n",
       "      <td>48</td>\n",
       "      <td>40</td>\n",
       "      <td>14648.90</td>\n",
       "      <td>0.026087</td>\n",
       "      <td>0.573913</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>MGTRLLCWAAICLLGADHTGAGVSQSLRHKVAKKGKDVALRYDPIS...</td>\n",
       "      <td>2698.266648</td>\n",
       "      <td>7949.531894</td>\n",
       "      <td>0.339425</td>\n",
       "      <td>0.191304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20383</th>\n",
       "      <td>A0A0A0MS05</td>\n",
       "      <td>114</td>\n",
       "      <td>45</td>\n",
       "      <td>36</td>\n",
       "      <td>14421.24</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.543860</td>\n",
       "      <td>0.429825</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>MGPGLLCWVLLCPLGEGPVDAGVTQSPTHLIKTRGQHVTLRCSPIS...</td>\n",
       "      <td>2802.882262</td>\n",
       "      <td>7582.409787</td>\n",
       "      <td>0.369656</td>\n",
       "      <td>0.175439</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20384 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id  length  hydr_count  polar_count  molecular_weight  \\\n",
       "0          Q8N7X0    1667         627          719         219717.24   \n",
       "1          Q5T1N1     836         243          401         107902.81   \n",
       "2          Q92667     903         340          335         113584.31   \n",
       "3          Q5VUY0     407         208          127          53468.44   \n",
       "4          P62736     377         161          143          48781.11   \n",
       "...           ...     ...         ...          ...               ...   \n",
       "20379      A0PK05     275         126           89          34825.86   \n",
       "20380      Q9HCN2     124          46           35          15149.80   \n",
       "20381  A0A0A0MS03     114          47           36          14631.45   \n",
       "20382  A0A0A6YYK4     115          48           40          14648.90   \n",
       "20383  A0A0A0MS05     114          45           36          14421.24   \n",
       "\n",
       "          helix      turn     sheet   A   C  ...    S    T   V   W   Y  \\\n",
       "0      0.225555  0.604079  0.170366  95  19  ...  154  104  95  28  49   \n",
       "1      0.183014  0.777512  0.039474  44  18  ...   91   55  24   3  18   \n",
       "2      0.079734  0.805094  0.115172  72  17  ...   94   49  72  10  16   \n",
       "3      0.491400  0.380835  0.127764  24  16  ...   23   14  39   8  13   \n",
       "4      0.445623  0.347480  0.206897  29   7  ...   26   24  20   4  16   \n",
       "...         ...       ...       ...  ..  ..  ...  ...  ...  ..  ..  ..   \n",
       "20379  0.316364  0.683636  0.000000  24   6  ...   23   23  15   3   7   \n",
       "20380  0.000000  0.750000  0.250000  11   2  ...   18    5   7   3   1   \n",
       "20381  0.026316  0.535088  0.438596   7   4  ...   12    4   5   2   5   \n",
       "20382  0.026087  0.573913  0.400000  12   3  ...   12    4   4   2   5   \n",
       "20383  0.026316  0.543860  0.429825   5   4  ...   12    5   6   1   5   \n",
       "\n",
       "                                          fasta_sequence  thsa_netsurfp2  \\\n",
       "0      MASKQTKKKEVHRINSAHGSDKSKDFYPFGSNVQSGSTEQKKGKFP...    25102.465221   \n",
       "1      MDEADFSEHTTYKQEDLPYDGDLSQIKIGNDYSFTSKKDGLEVLNQ...    14688.543593   \n",
       "2      MAIQFRSLFPLALPGMLALLGWWWFFSRKKGHVSSHDEQQVEAGAV...    19982.520280   \n",
       "3      MWDLALIFLAAACVFSLGVTLWVICSHFFTVHIPAAVGHPVKLRVL...     6416.179985   \n",
       "4      MCEEEDSTALVCDNGSGLCKAGFAGDDAPRAVFPSIVGRPRHQGVM...     4444.627391   \n",
       "...                                                  ...             ...   \n",
       "20379  MQLQVFWTGLEYTCRLLGITTAAVLIGVGTETFLQGQFKSLAFYLL...    10264.757559   \n",
       "20380  MGSSSEASFRSAQASCSGARRQGLGRGDQNLSVMPPNGRAQTHTPG...     2739.538957   \n",
       "20381  MGPGLLCWELLYLLGAGPVEAGVTQSPTHLIKTRGQQVTLRCSPIS...     2996.623152   \n",
       "20382  MGTRLLCWAAICLLGADHTGAGVSQSLRHKVAKKGKDVALRYDPIS...     2698.266648   \n",
       "20383  MGPGLLCWVLLCPLGEGPVDAGVTQSPTHLIKTRGQHVTLRCSPIS...     2802.882262   \n",
       "\n",
       "       tasa_netsurfp2  rhsa_netsurfp2  disorder  \n",
       "0       108276.162948        0.231837  0.335933  \n",
       "1        67203.514041        0.218568  0.533493  \n",
       "2        68025.600814        0.293750  0.644518  \n",
       "3        18254.852216        0.351478  0.000000  \n",
       "4        17613.271161        0.252345  0.037135  \n",
       "...               ...             ...       ...  \n",
       "20379    23347.907188        0.439644  0.483636  \n",
       "20380     8771.609425        0.312319  0.048387  \n",
       "20381     7889.445581        0.379827  0.184211  \n",
       "20382     7949.531894        0.339425  0.191304  \n",
       "20383     7582.409787        0.369656  0.175439  \n",
       "\n",
       "[20384 rows x 33 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exposed amino acid count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if RSA > 0.4 AA residue is exposed\n",
    "nsp_aa_exposed = Data_path + '/Raw/NSP_exposed_aa.csv'\n",
    "nsp_aa_exposed = pd.read_csv(nsp_aa_exposed, sep=',') \n",
    "df_features = pd.merge(df_features, nsp_aa_exposed, on=['id'], how=\"left\") \n",
    "# nps_aa_exposed contains only 19873 proteins, some get lost if inner merge is performed\n",
    "# now dataframe contains some NAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Solubility-weighted index (SWI)\n",
    "\n",
    "Bhandari BK, Gardner PP, Lim CS. **Solubility-Weighted Index: fast and accurate prediction of protein solubility.** Bioinformatics. 2020 Sep 15;36(18):4691-4698. doi: https://doi.org/10.1093/bioinformatics/btaa578"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "weights = {'A': 0.8356471476582918,\n",
    "           'C': 0.5208088354857734,\n",
    "           'U': 0.5208088354857734, # since FASTA sequences contain U \n",
    "           'E': 0.9876987431418378,\n",
    "           'D': 0.9079044671339564,\n",
    "           'G': 0.7997168496420723,\n",
    "           'F': 0.5849790194237692,\n",
    "           'I': 0.6784124413866582,\n",
    "           'H': 0.8947913996466419,\n",
    "           'K': 0.9267104557513497,\n",
    "           'M': 0.6296623675420369,\n",
    "           'L': 0.6554221515081433,\n",
    "           'N': 0.8597433107431216,\n",
    "           'Q': 0.789434648348208,\n",
    "           'P': 0.8235328714705341,\n",
    "           'S': 0.7440908318492778,\n",
    "           'R': 0.7712466317693457,\n",
    "           'T': 0.8096922697856334,\n",
    "           'W': 0.6374678690957594,\n",
    "           'V': 0.7357837119163659,\n",
    "           'Y': 0.6112801822947587}\n",
    "\n",
    "# Constants from logistic fitting\n",
    "# prob = 1 / (1 + exp(-(a * x + b)));\n",
    "\n",
    "A = 81.0581\n",
    "B = -62.7775\n",
    "\n",
    "def SWI(df):\n",
    "    df['SWI'] = df['fasta_sequence'].apply(\n",
    "        lambda x: np.mean([weights[i] for i in x]))\n",
    "    df['Prob. of Solubility'] = 1/(1 + np.exp(-(A*df['SWI'] + B)))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run SWI calculation, add SWI columns to feature dataset\n",
    "df_features = SWI(df_features.copy())\n",
    "# drop SWI column\n",
    "df_features = df_features.drop(columns=['SWI']) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggregation propensity\n",
    "\n",
    "Sánchez de Groot N, Pallarés I, Avilés FX, Vendrell J, Ventura S. **Prediction of \"hot spots\" of aggregation in disease-linked polypeptides.** BMC Struct Biol. 2005 Sep 30;5:18. doi: https://doi.org/10.1186/1472-6807-5-18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "propensities = {'I': 1.822,\n",
    "                'F': 1.754,\n",
    "                'V': 1.594,\n",
    "                'L': 1.380,\n",
    "                'Y': 1.159,\n",
    "                'W': 1.037,\n",
    "                'M': 0.910,\n",
    "                'C': 0.604,\n",
    "                'U': 0.604,\n",
    "                'A': -0.036,\n",
    "                'T': -0.159,\n",
    "                'S': -0.294,\n",
    "                'P': -0.334,\n",
    "                'G': -0.535,\n",
    "                'K': -0.931,\n",
    "                'H': -1.033,\n",
    "                'Q': -1.231,\n",
    "                'R': -1.240,\n",
    "                'N': -1.302,\n",
    "                'E': -1.412,\n",
    "                'D': -1.836}\n",
    "\n",
    "def agg_score(df):\n",
    "    df['Aggregation_propensity'] = df['fasta_sequence'].apply(\n",
    "        lambda x: np.sum([propensities[i] for i in x]))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run aggregation propensity calculation, add aggregation column to feature dataset\n",
    "df_features = agg_score(df_features.copy())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simple protein analysis by Biopython\n",
    "\n",
    "Bio.SeqUtils.ProtParam module (https://biopython.org/wiki/ProtParam)\n",
    "- Aromaticity: Calculates the aromaticity value of a protein according to Lobry, 1994. It is simply the relative frequency of Phe+Trp+Tyr.\n",
    "- Instability index: Implementation of the method of Guruprasad et al. 1990 to test a protein for stability. Any value above 40 means the protein is unstable (has a short half life). \n",
    "- Gravy: Calculate the gravy according to Kyte and Doolittle.\n",
    "- Isoelectric point: Calculate the isoelectric point.\n",
    "- Charge at pH 5: Calculate the charge of a protein at given pH.\n",
    "- Charge at pH 7: Calculate the charge of a protein at given pH."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a list and replace U with C (Biopython gives an error otherwise)\n",
    "sequences = list(df_features['fasta_sequence'])\n",
    "sequences = [sequence.replace('U', 'C') for sequence in sequences]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create empty lists\n",
    "aromaticity_list = []\n",
    "instability_index_list = []\n",
    "gravy_list = []\n",
    "isoelectric_point_list = []\n",
    "charge_at_7_list = []\n",
    "charge_at_5_list = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sequence in sequences:\n",
    "    analysis = ProteinAnalysis(str(sequence))\n",
    "    \n",
    "    #calculate different values\n",
    "    aromaticity = analysis.aromaticity()\n",
    "    instability_index = analysis.instability_index()\n",
    "    gravy_score = analysis.gravy()\n",
    "    isoelectric_point = analysis.isoelectric_point()\n",
    "    charge_at_7 = analysis.charge_at_pH(7)\n",
    "    charge_at_5 = analysis.charge_at_pH(5)\n",
    "    \n",
    "    #append the list\n",
    "    aromaticity_list.append(aromaticity)\n",
    "    instability_index_list.append(instability_index)\n",
    "    gravy_list.append(gravy_score)\n",
    "    isoelectric_point_list.append(isoelectric_point)\n",
    "    charge_at_7_list.append(charge_at_7)\n",
    "    charge_at_5_list.append(charge_at_5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a column\n",
    "df_features['Aromaticity'] = aromaticity_list \n",
    "df_features['Instability_index'] = instability_index_list \n",
    "df_features['Gravy'] = gravy_list \n",
    "df_features['isoelectric_point'] = isoelectric_point_list \n",
    "df_features['charge_at_7'] = charge_at_7_list \n",
    "df_features['charge_at_5'] = charge_at_5_list "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HSP annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "HSP = Data_path + '/Raw/UniProt/HSP.tab'\n",
    "HSP = pd.read_csv(HSP, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "HSP_list = set(HSP[\"Entry\"])\n",
    "df_features[\"HSP\"] = np.where(df_features['id'].isin(HSP_list), 1, 0) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Post-translational modification"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### UniProt annotations (keywords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def up_import(ptm):\n",
    "    UP_list = []\n",
    "    f = open(Data_path + '/Raw/UniProt/' + ptm + '.list', 'r')\n",
    "    \n",
    "    # import data\n",
    "    while True:\n",
    "        line = f.readline()\n",
    "        line = line.rstrip(\"\\n\")\n",
    "        UP_list.append(line) \n",
    "        # end of file is reached\n",
    "        if not line:\n",
    "            break\n",
    "    \n",
    "    return UP_list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define UniProt keywords\n",
    "UP = [\"PTM\", \"Acetylation\", \"Citrullination\", \"Glycoprotein\", \"GPI-anchor\", \"Lipoprotein\", \"Methylation\",  \"Myristate\", \"Nitration\", \"Palmitate\", \n",
    "      \"Prenylation\", \"Phosphoprotein\", \"S-nitrosylation\", \"Ubl_conjugation\"]\n",
    "# create dict of all PTM files\n",
    "UP_dict = {}\n",
    "\n",
    "for i in UP:\n",
    "    UP_dict[i] = up_import(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of annotations in feature dataset (UniProt keywords)\n",
      "PTM: 13918\n",
      "Citrullination: 68\n",
      "GPI-anchor: 139\n",
      "Lipoprotein: 866\n",
      "Nitration: 48\n",
      "Prenylation: 172\n"
     ]
    }
   ],
   "source": [
    "# 0/1 annotation\n",
    "UP_only = [\"PTM\", \"Citrullination\", \"GPI-anchor\", \"Lipoprotein\", \"Nitration\", \"Prenylation\"]\n",
    "\n",
    "print(\"Number of annotations in feature dataset (UniProt keywords)\")\n",
    "for i in UP_only:\n",
    "    column_name = str(i) + \"_UP\"\n",
    "    df_features[column_name] = np.where(df_features['id'].isin(UP_dict[i]), 1, 0) \n",
    "    print(str(i) + \":\", sum(df_features[column_name]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### UniProt annotations (text mining)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load PTM comment table\n",
    "PTM_comments = Data_path + '/Raw/UniProt/PTM_comments.tab'\n",
    "PTM_comments = pd.read_csv(PTM_comments, sep='\\t', engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge PTM comment table with feature dataset\n",
    "PTM_comments.rename(columns = {'Entry':'id', 'Post-translational modification':'PTM_comments'}, inplace = True)\n",
    "df_features = df_features.merge(PTM_comments, on='id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of annotations in feature dataset (UniProt text mining)\n",
      "ISGylation: 42\n",
      "NEDDylation: 41\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of annotations in feature dataset (UniProt text mining)\")\n",
    "\n",
    "df_features['ISGylation_UP'] = np.where(df_features['PTM_comments'].str.contains(\"ISGy\",na=False), 1, 0)\n",
    "print(\"ISGylation:\", sum(df_features['ISGylation_UP']))\n",
    "df_features['NEDDylation_UP'] = np.where(df_features['PTM_comments'].str.contains(\"NEDD\",na=False), 1, 0)\n",
    "print(\"NEDDylation:\", sum(df_features['NEDDylation_UP']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PhosphoSitePlus annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def psp_import(ptm):\n",
    "    # import data\n",
    "    loc = Data_path + '/Raw/PhosphositePlus/' + ptm + '_psp.tab'\n",
    "    df = pd.read_csv(loc, sep='\\t', engine='python')\n",
    "    # filter for human entries only\n",
    "    df = df[df[\"ORGANISM\"] == \"human\"]\n",
    "    return df \n",
    "    \n",
    "def psp_to_list(df):    \n",
    "    # create list of UniProt IDs\n",
    "    UP_list = list(df[\"ACC_ID\"])\n",
    "    print(\"\\t Number of entries:\", len(UP_list))\n",
    "    # filter out duplicates\n",
    "    UP_list_unique = list(set(UP_list))\n",
    "    print(\"\\t Number of unique entries:\", len(UP_list_unique))\n",
    "    return UP_list_unique\n",
    "\n",
    "def psp_to_count(df):    \n",
    "    # count entries per UniProt ID\n",
    "    df[\"COUNT\"] = df.groupby(\"ACC_ID\")[\"ACC_ID\"].transform('count')\n",
    "    df = df.drop_duplicates(subset=[\"ACC_ID\"], keep='first')\n",
    "    return df[[\"ACC_ID\", \"COUNT\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ubiquitination (PhosphositePlus)\n",
      "\t Number of entries: 97777\n",
      "\t Number of unique entries: 12435\n",
      "Phosphorylation (PhosphositePlus)\n",
      "\t Number of entries: 239544\n",
      "\t Number of unique entries: 19834\n",
      "Methylation (PhosphositePlus)\n",
      "\t Number of entries: 16374\n",
      "\t Number of unique entries: 5691\n",
      "Acetylation (PhosphositePlus)\n",
      "\t Number of entries: 22824\n",
      "\t Number of unique entries: 7253\n",
      "Sumoylation (PhosphositePlus)\n",
      "\t Number of entries: 8415\n",
      "\t Number of unique entries: 2667\n",
      "O-GlcNAc (PhosphositePlus)\n",
      "\t Number of entries: 441\n",
      "\t Number of unique entries: 173\n",
      "O-GalNAc (PhosphositePlus)\n",
      "\t Number of entries: 2102\n",
      "\t Number of unique entries: 478\n"
     ]
    }
   ],
   "source": [
    "psp = [\"Ubiquitination\", \"Phosphorylation\", \"Methylation\", \"Acetylation\", \"Sumoylation\", \"O-GlcNAc\", \"O-GalNAc\"]\n",
    "# create dict of all PTM files\n",
    "psp_dict = {}\n",
    "\n",
    "for i in psp:\n",
    "    print(i, \"(PhosphositePlus)\")\n",
    "    psp_dict[i] = psp_to_list(psp_import(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create Glycosylation column by combining O-GlcNAc and O-GalNAc\n",
    "Glycosylation_PSP = list(set(psp_dict[\"O-GlcNAc\"] + psp_dict[\"O-GalNAc\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### iPTMnet annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "iPTM = Data_path + '/Raw/iPTMnet_annotations.txt'  \n",
    "iPTM = pd.read_csv(iPTM, sep='\\t', engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "iPTM = iPTM[iPTM[\"organism\"] == \"Homo sapiens (Human)\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# count all glycosylation sites together\n",
    "iPTM[\"ptm_type\"] = iPTM[\"ptm_type\"].replace([\"N-GLYCOSYLATION\", \"O-GLYCOSYLATION\", \"C-GLYCOSYLATION\", \"S-GLYCOSYLATION\"], \n",
    "                    \"GLYCOSYLATION\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "iPTM_list = [\"PHOSPHORYLATION\", \"ACETYLATION\", \"GLYCOSYLATION\", \"S-NITROSYLATION\", \"METHYLATION\", \"UBIQUITINATION\", \n",
    "        \"MYRISTOYLATION\", \"SUMOYLATION\"] \n",
    "\n",
    "def iPTM_to_list(ptm): \n",
    "    \n",
    "    iPTM_subset = iPTM[iPTM[\"ptm_type\"] == ptm]\n",
    "    print(\"\\t Number of entries:\", len(iPTM_subset))\n",
    "    \n",
    "    UP_list = list(set(list(iPTM_subset[\"substrate_UniProtAC\"])))\n",
    "    print(\"\\t Number of unique entries:\", len(UP_list))\n",
    "    \n",
    "    return UP_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PHOSPHORYLATION (iPTMnet)\n",
      "\t Number of entries: 139966\n",
      "\t Number of unique entries: 15228\n",
      "ACETYLATION (iPTMnet)\n",
      "\t Number of entries: 4841\n",
      "\t Number of unique entries: 2970\n",
      "GLYCOSYLATION (iPTMnet)\n",
      "\t Number of entries: 3348\n",
      "\t Number of unique entries: 1160\n",
      "S-NITROSYLATION (iPTMnet)\n",
      "\t Number of entries: 2376\n",
      "\t Number of unique entries: 1213\n",
      "METHYLATION (iPTMnet)\n",
      "\t Number of entries: 1007\n",
      "\t Number of unique entries: 471\n",
      "UBIQUITINATION (iPTMnet)\n",
      "\t Number of entries: 523\n",
      "\t Number of unique entries: 183\n",
      "MYRISTOYLATION (iPTMnet)\n",
      "\t Number of entries: 125\n",
      "\t Number of unique entries: 120\n",
      "SUMOYLATION (iPTMnet)\n",
      "\t Number of entries: 35\n",
      "\t Number of unique entries: 25\n"
     ]
    }
   ],
   "source": [
    "# create dict of all PTM files\n",
    "iPTM_dict = {}\n",
    "\n",
    "for i in iPTM_list:\n",
    "    print(i, \"(iPTMnet)\")\n",
    "    iPTM_dict[i] = iPTM_to_list(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SwissPalm (Palmitoylation database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "SwissPalm = Data_path + '/Raw/SwissPalm_proteins.txt'\n",
    "SwissPalm = pd.read_csv(SwissPalm, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sp_to_list(df): \n",
    "    # filter for human entries only\n",
    "    df = df[df[\"organism\"] == \"Homo sapiens\"]\n",
    "    # create list of UniProt IDs\n",
    "    UP_list = list(df[\"uniprot_ac\"]) \n",
    "    # filter out duplicates\n",
    "    UP_list_unique = list(set(UP_list))\n",
    "    \n",
    "    return UP_list_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique entries in SwissPalm: 4587\n"
     ]
    }
   ],
   "source": [
    "SwissPalm_list = sp_to_list(SwissPalm)\n",
    "print(\"Number of unique entries in SwissPalm:\", len(SwissPalm_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combined PTM annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create list of combined database annotations\n",
    "\n",
    "Acetylation_all = set(UP_dict[\"Acetylation\"] + iPTM_dict[\"ACETYLATION\"] + psp_dict[\"Acetylation\"])\n",
    "Glycosylation_all = set(UP_dict[\"Glycoprotein\"] + iPTM_dict[\"GLYCOSYLATION\"] + Glycosylation_PSP)\n",
    "Methylation_all = set(UP_dict[\"Methylation\"] + iPTM_dict[\"METHYLATION\"] + psp_dict[\"Methylation\"])\n",
    "Myristoylation_all = set(UP_dict[\"Myristate\"] + iPTM_dict[\"MYRISTOYLATION\"])\n",
    "Nitrosylation_all = set(UP_dict[\"S-nitrosylation\"] + iPTM_dict[\"S-NITROSYLATION\"])                  \n",
    "Palmitoylation_all = set(UP_dict[\"Palmitate\"] + SwissPalm_list)\n",
    "Phosphorylation_all = set(UP_dict[\"Phosphoprotein\"] + iPTM_dict[\"PHOSPHORYLATION\"] + psp_dict[\"Phosphorylation\"])\n",
    "SUMOylation_all = set(iPTM_dict[\"SUMOYLATION\"] + psp_dict[\"Sumoylation\"])\n",
    "Ubiquitination_all = set(iPTM_dict[\"UBIQUITINATION\"] + psp_dict[\"Ubiquitination\"])\n",
    "\n",
    "df_features['Acetylation_all'] = np.where(df_features['id'].isin(Acetylation_all), 1, 0)\n",
    "df_features['Glycosylation_all'] = np.where(df_features['id'].isin(Glycosylation_all), 1, 0)\n",
    "df_features['Methylation_all'] = np.where(df_features['id'].isin(Methylation_all), 1, 0)\n",
    "df_features['Myristoylation_all'] = np.where(df_features['id'].isin(Myristoylation_all), 1, 0)\n",
    "df_features['Nitrosylation_all'] = np.where(df_features['id'].isin(Nitrosylation_all), 1, 0)\n",
    "df_features['Palmitoylation_all'] = np.where(df_features['id'].isin(Palmitoylation_all), 1, 0)\n",
    "df_features['Phosphorylation_all'] = np.where(df_features['id'].isin(Phosphorylation_all), 1, 0)\n",
    "df_features['SUMOylation_all'] = np.where(df_features['id'].isin(SUMOylation_all), 1, 0)\n",
    "df_features['Ubiquitination_all'] = np.where(df_features['id'].isin(Ubiquitination_all), 1, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of combined annotations in feature dataset\n",
      "Acetylation: 7950\n",
      "Glycosylation: 4892\n",
      "Methylation: 5683\n",
      "Myristoylation: 196\n",
      "Nitrosylation: 1223\n",
      "Palmitoylation: 3705\n",
      "Phosphorylation: 17524\n",
      "Palmitoylation: 3705\n",
      "Ubiquitination: 11709\n"
     ]
    }
   ],
   "source": [
    "# number of annotations\n",
    "print(\"Number of combined annotations in feature dataset\")\n",
    "print(\"Acetylation:\", sum(df_features['Acetylation_all']))\n",
    "print(\"Glycosylation:\", sum(df_features['Glycosylation_all']))\n",
    "print(\"Methylation:\", sum(df_features['Methylation_all']))\n",
    "print(\"Myristoylation:\", sum(df_features['Myristoylation_all']))\n",
    "print(\"Nitrosylation:\", sum(df_features['Nitrosylation_all']))\n",
    "print(\"Palmitoylation:\", sum(df_features['Palmitoylation_all']))\n",
    "print(\"Phosphorylation:\", sum(df_features['Phosphorylation_all']))\n",
    "print(\"Palmitoylation:\", sum(df_features['Palmitoylation_all']))\n",
    "print(\"Ubiquitination:\", sum(df_features['Ubiquitination_all']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MusiteDeep (PTM prediction)\n",
    "\n",
    "Wang D, Liu D, Yuchi J, He F, Jiang Y, Cai S, Li J, Xu D. **MusiteDeep: a deep-learning based webserver for protein post-translational modification site prediction and visualization.** Nucleic Acids Res. 2020 Jul 2;48(W1):W140-W146. doi: https://doi.org/10.1093/nar/gkaa275"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load PTM prediction dataset\n",
    "MSP = Data_path + '/Raw/MusiteDeep/PTM_MusiteDeep_75.csv'\n",
    "MSP = pd.read_csv(MSP, sep=',', engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge PTM predictions with feature dataset\n",
    "df_features = df_features.merge(MSP, how=\"left\", left_on=\"id\", right_on=\"ID\")\n",
    "df_features = df_features.drop(columns=[\"ID\"])\n",
    "\n",
    "# change any NA values to 0\n",
    "df_features = df_features.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert PTM count to 0/1 annotation\n",
    "df_features[\"PTM_count\"] = np.where(df_features[\"PTM_count\"] > 0, 1, 0)\n",
    "# rename columns\n",
    "df_features.rename(columns = {'PTM_count':'PTM_MSD', 'Acetyllysine_MSD':'Acetylation_MSD'}, inplace=True)\n",
    "# drop MSD predictions with no annotation \n",
    "df_features = df_features.drop(columns=['Hydroxylation_MSD', 'Pyrrolidone_carboxylic_acid_MSD']) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PROSITE domains\n",
    "\n",
    "#### Coiled coil domain\n",
    "- PS00029 (LEUCINE_ZIPPER)\n",
    "- PS00914 (SYNTAXIN)\n",
    "- PS50027 (EGF_LAM_2)\n",
    "- PS50067 (KINESIN_MOTOR_2)\n",
    "- PS50119 (ZF_BBOX)\n",
    "- PS50151 (UVR)\n",
    "- PS51121 (NTA)\n",
    "- PS51131 (ZN_HOOK)\n",
    "- PS51338 (IMD)\n",
    "- PS51370 (R)\n",
    "- PS51376 (DBB)\n",
    "- PS51386 (RINT1_TIP20)\n",
    "- PS51649 (NPH3)\n",
    "- PS51663 (STATHMIN_3)\n",
    "- PS51776 (RH1)\n",
    "- PS51842 (IF_ROD_2)\n",
    "\n",
    "#### EGF\n",
    "- PS50026 (EGF_3)\n",
    "- PS00022 (EGF_1)\n",
    "- PS01186 (EGF_2)\n",
    "\n",
    "#### RAS profile\n",
    "- PS51421 (RAS) \n",
    "\n",
    "#### RRM\n",
    "- PS50102 (RRM)\n",
    "\n",
    "#### WW domain\n",
    "- PS01179 (PID)\n",
    "- PS50020 (WW_DOMAIN_2)\n",
    "- PS51035 (BAG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROSITE_path = Data_path + '/Raw/Prosite/Matched/'\n",
    "files_PROSITE = [f for f in os.listdir(PROSITE_path) if isfile(join(PROSITE_path, f))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "for file in files_PROSITE[0:]:\n",
    "    domain = ''.join(file.split('.')[:-1])\n",
    "    file = pd.read_csv(PROSITE_path + file, sep='\\t', engine='python', header=None)\n",
    "    domain_set = set(file[0])\n",
    "    df_features[domain] = np.where(df_features['id'].isin(domain_set), 1, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transmembrane annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data (tm=transmembrane)\n",
    "tm = Data_path + '/Raw/UniProt/TM_proteins.tab'\n",
    "tm = pd.read_csv(tm, sep='\\t')\n",
    "tm_proteins = list(tm['Entry'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of transmembrane proteins in feature dataset: 6316\n"
     ]
    }
   ],
   "source": [
    "# add a column 1/0\n",
    "df_features['transmembrane'] = np.where(df_features['id'].isin(tm_proteins), 1, 0)\n",
    "print(\"Number of transmembrane proteins in feature dataset:\", sum(df_features['transmembrane']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transmembrane predictions (TMHMM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data (tm=transmembrane)\n",
    "TMHMM = Data_path + '/Raw/TMHMM_tmp.csv'\n",
    "TMHMM = pd.read_csv(TMHMM, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of transmembrane proteins predicted by TMHMM: 5279\n"
     ]
    }
   ],
   "source": [
    "TMHMM_list = list(TMHMM['id'])\n",
    "# add a column 1/0\n",
    "df_features['TMHMM'] = np.where(df_features['id'].isin(TMHMM_list), 1, 0)\n",
    "print(\"Number of transmembrane proteins predicted by TMHMM:\", sum(df_features['TMHMM']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data preprocessing "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_normalized = df_features.copy()\n",
    "\n",
    "# drop ununsed column\n",
    "df_normalized = df_normalized.drop(columns=['PTM_comments']) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Previous normalisation way\n",
    "\n",
    "features_count = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y',\n",
    "                  'hydr_count', 'polar_count']\n",
    "features_exposed = ['A_exposed', 'C_exposed', 'D_exposed', 'E_exposed', 'F_exposed', 'G_exposed', 'H_exposed', 'I_exposed', \n",
    "                    'K_exposed', 'L_exposed', 'M_exposed', 'N_exposed', 'P_exposed', 'Q_exposed', 'R_exposed', 'S_exposed', \n",
    "                    'T_exposed', 'V_exposed', 'W_exposed', 'Y_exposed']\n",
    "hydrophobic = ['A_exposed', 'C_exposed', 'F_exposed', 'L_exposed', 'I_exposed', 'M_exposed', 'V_exposed', 'W_exposed', \n",
    "               'Y_exposed']\n",
    "polar = ['D_exposed', 'E_exposed', 'H_exposed', 'K_exposed', 'N_exposed', 'Q_exposed', 'R_exposed', 'T_exposed']\n",
    "\n",
    "# create a sum of exposed AA column\n",
    "df_normalized['Sum_AA_exposed'] = df_normalized[features_exposed].sum(axis=1)\n",
    "\n",
    "# create a column for polar and hydrophobic exposed AA counts and normalise it\n",
    "df_normalized['Polar_exposed'] = df_normalized[polar].sum(axis=1)\n",
    "df_normalized['Polar_exposed'] = df_normalized['Polar_exposed']/df_normalized['Sum_AA_exposed']\n",
    "\n",
    "df_normalized['Hydrophobic_exposed'] = df_normalized[hydrophobic].sum(axis=1)\n",
    "df_normalized['Hydrophobic_exposed'] = df_normalized['Hydrophobic_exposed']/df_normalized['Sum_AA_exposed']\n",
    "\n",
    "for i, feature in enumerate(features_exposed):\n",
    "    df_normalized[feature] = df_normalized[feature]/df_normalized['Sum_AA_exposed']\n",
    "\n",
    "# normalize amino acid count features by length\n",
    "for feature in features_count:\n",
    "    df_normalized[feature] = df_normalized[feature]/df_normalized.length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# handle infinity values as missing values\n",
    "pd.options.mode.use_inf_as_na = True\n",
    "# change any NA values to 0, caused by division by 0\n",
    "df_normalized = df_normalized.fillna(0)\n",
    "# drop the summed column\n",
    "df_normalized = df_normalized.drop(columns=['Sum_AA_exposed'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A_exposed</th>\n",
       "      <th>C_exposed</th>\n",
       "      <th>D_exposed</th>\n",
       "      <th>E_exposed</th>\n",
       "      <th>F_exposed</th>\n",
       "      <th>G_exposed</th>\n",
       "      <th>H_exposed</th>\n",
       "      <th>I_exposed</th>\n",
       "      <th>K_exposed</th>\n",
       "      <th>L_exposed</th>\n",
       "      <th>M_exposed</th>\n",
       "      <th>N_exposed</th>\n",
       "      <th>P_exposed</th>\n",
       "      <th>Q_exposed</th>\n",
       "      <th>R_exposed</th>\n",
       "      <th>S_exposed</th>\n",
       "      <th>T_exposed</th>\n",
       "      <th>V_exposed</th>\n",
       "      <th>W_exposed</th>\n",
       "      <th>Y_exposed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.062548</td>\n",
       "      <td>0.006796</td>\n",
       "      <td>0.068159</td>\n",
       "      <td>0.106898</td>\n",
       "      <td>0.016816</td>\n",
       "      <td>0.074295</td>\n",
       "      <td>0.025235</td>\n",
       "      <td>0.015650</td>\n",
       "      <td>0.089634</td>\n",
       "      <td>0.049017</td>\n",
       "      <td>0.016064</td>\n",
       "      <td>0.047195</td>\n",
       "      <td>0.078232</td>\n",
       "      <td>0.065021</td>\n",
       "      <td>0.079960</td>\n",
       "      <td>0.093955</td>\n",
       "      <td>0.054650</td>\n",
       "      <td>0.029749</td>\n",
       "      <td>0.006630</td>\n",
       "      <td>0.013497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.032369</td>\n",
       "      <td>0.013637</td>\n",
       "      <td>0.030665</td>\n",
       "      <td>0.045436</td>\n",
       "      <td>0.019845</td>\n",
       "      <td>0.036529</td>\n",
       "      <td>0.016012</td>\n",
       "      <td>0.016973</td>\n",
       "      <td>0.046236</td>\n",
       "      <td>0.035278</td>\n",
       "      <td>0.010941</td>\n",
       "      <td>0.025764</td>\n",
       "      <td>0.040681</td>\n",
       "      <td>0.028245</td>\n",
       "      <td>0.034466</td>\n",
       "      <td>0.037302</td>\n",
       "      <td>0.024926</td>\n",
       "      <td>0.019336</td>\n",
       "      <td>0.009682</td>\n",
       "      <td>0.015842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.040293</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.046646</td>\n",
       "      <td>0.075188</td>\n",
       "      <td>0.004545</td>\n",
       "      <td>0.051242</td>\n",
       "      <td>0.015075</td>\n",
       "      <td>0.004785</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.023933</td>\n",
       "      <td>0.008728</td>\n",
       "      <td>0.028160</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>0.056842</td>\n",
       "      <td>0.069652</td>\n",
       "      <td>0.039526</td>\n",
       "      <td>0.016605</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.057471</td>\n",
       "      <td>0.004115</td>\n",
       "      <td>0.066327</td>\n",
       "      <td>0.105462</td>\n",
       "      <td>0.010680</td>\n",
       "      <td>0.069592</td>\n",
       "      <td>0.023392</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.084507</td>\n",
       "      <td>0.041150</td>\n",
       "      <td>0.013333</td>\n",
       "      <td>0.044944</td>\n",
       "      <td>0.071259</td>\n",
       "      <td>0.062745</td>\n",
       "      <td>0.075758</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.052239</td>\n",
       "      <td>0.026667</td>\n",
       "      <td>0.003257</td>\n",
       "      <td>0.009967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.079208</td>\n",
       "      <td>0.009149</td>\n",
       "      <td>0.087791</td>\n",
       "      <td>0.136646</td>\n",
       "      <td>0.021568</td>\n",
       "      <td>0.091429</td>\n",
       "      <td>0.032866</td>\n",
       "      <td>0.021053</td>\n",
       "      <td>0.118280</td>\n",
       "      <td>0.065434</td>\n",
       "      <td>0.020134</td>\n",
       "      <td>0.063218</td>\n",
       "      <td>0.098886</td>\n",
       "      <td>0.080463</td>\n",
       "      <td>0.098160</td>\n",
       "      <td>0.114054</td>\n",
       "      <td>0.066483</td>\n",
       "      <td>0.038889</td>\n",
       "      <td>0.009346</td>\n",
       "      <td>0.018018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.349515</td>\n",
       "      <td>0.262712</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.345679</td>\n",
       "      <td>0.380952</td>\n",
       "      <td>0.152174</td>\n",
       "      <td>0.205882</td>\n",
       "      <td>0.397260</td>\n",
       "      <td>0.987500</td>\n",
       "      <td>0.480000</td>\n",
       "      <td>0.670543</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.176471</td>\n",
       "      <td>0.246377</td>\n",
       "      <td>0.432432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          A_exposed     C_exposed     D_exposed     E_exposed     F_exposed  \\\n",
       "count  20384.000000  20384.000000  20384.000000  20384.000000  20384.000000   \n",
       "mean       0.062548      0.006796      0.068159      0.106898      0.016816   \n",
       "std        0.032369      0.013637      0.030665      0.045436      0.019845   \n",
       "min        0.000000      0.000000      0.000000      0.000000      0.000000   \n",
       "25%        0.040293      0.000000      0.046646      0.075188      0.004545   \n",
       "50%        0.057471      0.004115      0.066327      0.105462      0.010680   \n",
       "75%        0.079208      0.009149      0.087791      0.136646      0.021568   \n",
       "max        0.333333      0.349515      0.262712      0.500000      0.217391   \n",
       "\n",
       "          G_exposed     H_exposed     I_exposed     K_exposed     L_exposed  \\\n",
       "count  20384.000000  20384.000000  20384.000000  20384.000000  20384.000000   \n",
       "mean       0.074295      0.025235      0.015650      0.089634      0.049017   \n",
       "std        0.036529      0.016012      0.016973      0.046236      0.035278   \n",
       "min        0.000000      0.000000      0.000000      0.000000      0.000000   \n",
       "25%        0.051242      0.015075      0.004785      0.055556      0.023933   \n",
       "50%        0.069592      0.023392      0.011364      0.084507      0.041150   \n",
       "75%        0.091429      0.032866      0.021053      0.118280      0.065434   \n",
       "max        0.750000      0.300000      0.500000      0.345679      0.380952   \n",
       "\n",
       "          M_exposed     N_exposed     P_exposed     Q_exposed     R_exposed  \\\n",
       "count  20384.000000  20384.000000  20384.000000  20384.000000  20384.000000   \n",
       "mean       0.016064      0.047195      0.078232      0.065021      0.079960   \n",
       "std        0.010941      0.025764      0.040681      0.028245      0.034466   \n",
       "min        0.000000      0.000000      0.000000      0.000000      0.000000   \n",
       "25%        0.008728      0.028160      0.050000      0.046875      0.056842   \n",
       "50%        0.013333      0.044944      0.071259      0.062745      0.075758   \n",
       "75%        0.020134      0.063218      0.098886      0.080463      0.098160   \n",
       "max        0.152174      0.205882      0.397260      0.987500      0.480000   \n",
       "\n",
       "          S_exposed     T_exposed     V_exposed     W_exposed     Y_exposed  \n",
       "count  20384.000000  20384.000000  20384.000000  20384.000000  20384.000000  \n",
       "mean       0.093955      0.054650      0.029749      0.006630      0.013497  \n",
       "std        0.037302      0.024926      0.019336      0.009682      0.015842  \n",
       "min        0.000000      0.000000      0.000000      0.000000      0.000000  \n",
       "25%        0.069652      0.039526      0.016605      0.000000      0.004673  \n",
       "50%        0.090909      0.052239      0.026667      0.003257      0.009967  \n",
       "75%        0.114054      0.066483      0.038889      0.009346      0.018018  \n",
       "max        0.670543      0.600000      0.176471      0.246377      0.432432  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_normalized[features_exposed].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Log transformation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_log = df_normalized.copy()\n",
    "df_log[\"length\"] = df_log[\"length\"].transform(np.log2)\n",
    "df_log[\"molecular_weight\"] = df_log[\"molecular_weight\"].transform(np.log2)\n",
    "df_log[\"tasa_netsurfp2\"] = df_log[\"tasa_netsurfp2\"].transform(np.log2)\n",
    "df_log[\"thsa_netsurfp2\"] = df_log[\"thsa_netsurfp2\"].transform(np.log2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>tasa_netsurfp2</th>\n",
       "      <th>thsa_netsurfp2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20384.000000</td>\n",
       "      <td>20381.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>8.668189</td>\n",
       "      <td>15.684163</td>\n",
       "      <td>14.681780</td>\n",
       "      <td>12.842945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.137330</td>\n",
       "      <td>1.137987</td>\n",
       "      <td>1.057258</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>8.120497</td>\n",
       "      <td>8.356479</td>\n",
       "      <td>7.222158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>7.960002</td>\n",
       "      <td>14.977911</td>\n",
       "      <td>13.996037</td>\n",
       "      <td>12.198453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>8.696968</td>\n",
       "      <td>15.711864</td>\n",
       "      <td>14.620270</td>\n",
       "      <td>12.853003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>9.388017</td>\n",
       "      <td>16.406296</td>\n",
       "      <td>15.362132</td>\n",
       "      <td>13.440654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>13.824462</td>\n",
       "      <td>20.763749</td>\n",
       "      <td>19.447881</td>\n",
       "      <td>17.275738</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             length  molecular_weight  tasa_netsurfp2  thsa_netsurfp2\n",
       "count  20384.000000      20384.000000    20384.000000    20381.000000\n",
       "mean       8.668189         15.684163       14.681780       12.842945\n",
       "std        1.137330          1.137987        1.057258             NaN\n",
       "min        1.000000          8.120497        8.356479        7.222158\n",
       "25%        7.960002         14.977911       13.996037       12.198453\n",
       "50%        8.696968         15.711864       14.620270       12.853003\n",
       "75%        9.388017         16.406296       15.362132       13.440654\n",
       "max       13.824462         20.763749       19.447881       17.275738"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_log[[\"length\", \"molecular_weight\", \"tasa_netsurfp2\", \"thsa_netsurfp2\"]].describe() # std is NaN after log transformation?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_log = df_log.dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Removal of MS bias\n",
    "\n",
    "According to ProteomicsDB (https://www.proteomicsdb.org/#api), the following evidence classes exist:\n",
    "- 0: not observed\n",
    "- 1: observed with low scoring spectra \n",
    "- 2: observed with high quality spectra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load evidence ProteomicsDB data\n",
    "with open(Data_path + \"/Raw/ProteomicsDB/ProteomicsDB_evidence_positive.txt\") as f:\n",
    "    ProteomicsDB_evidence_positive = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins in positive set: 18997\n",
      "Number of proteins in positive set with evidence score of 1 or higher (i.e., MS-detectable): 16791\n"
     ]
    }
   ],
   "source": [
    "# create list of proteins with evidence on ProteomicsDB\n",
    "pos = set(ProteomicsDB_evidence_positive.keys())\n",
    "pos_filtered_low = set([k for k, v in ProteomicsDB_evidence_positive.items() if v > 0])\n",
    "print(\"Number of proteins in positive set:\", len(pos))\n",
    "print(\"Number of proteins in positive set with evidence score of 1 or higher (i.e., MS-detectable):\", len(pos_filtered_low))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of MS-detectable proteins in human proteome dataset: 16790\n"
     ]
    }
   ],
   "source": [
    "# filter feature dataset for MS detected proteins\n",
    "df_MS_filtered = df_log[df_log[\"id\"].isin(pos_filtered_low)]\n",
    "\n",
    "print(\"Number of MS-detectable proteins in human proteome dataset:\", len(df_MS_filtered))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save final feature data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>length</th>\n",
       "      <th>hydr_count</th>\n",
       "      <th>polar_count</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>helix</th>\n",
       "      <th>turn</th>\n",
       "      <th>sheet</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>...</th>\n",
       "      <th>Methylation_MSD</th>\n",
       "      <th>coiled_coil</th>\n",
       "      <th>EGF</th>\n",
       "      <th>RAS_profile</th>\n",
       "      <th>RRM</th>\n",
       "      <th>ww_domain</th>\n",
       "      <th>transmembrane</th>\n",
       "      <th>TMHMM</th>\n",
       "      <th>Polar_exposed</th>\n",
       "      <th>Hydrophobic_exposed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q92667</td>\n",
       "      <td>9.818582</td>\n",
       "      <td>0.376523</td>\n",
       "      <td>0.370986</td>\n",
       "      <td>16.793404</td>\n",
       "      <td>0.079734</td>\n",
       "      <td>0.805094</td>\n",
       "      <td>0.115172</td>\n",
       "      <td>0.079734</td>\n",
       "      <td>0.018826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.447738</td>\n",
       "      <td>0.263651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P62736</td>\n",
       "      <td>8.558421</td>\n",
       "      <td>0.427056</td>\n",
       "      <td>0.379310</td>\n",
       "      <td>15.574035</td>\n",
       "      <td>0.445623</td>\n",
       "      <td>0.347480</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.018568</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.682540</td>\n",
       "      <td>0.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q9H553</td>\n",
       "      <td>8.700440</td>\n",
       "      <td>0.471154</td>\n",
       "      <td>0.358173</td>\n",
       "      <td>15.735720</td>\n",
       "      <td>0.485577</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.139423</td>\n",
       "      <td>0.072115</td>\n",
       "      <td>0.028846</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.614286</td>\n",
       "      <td>0.185714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P0C7M7</td>\n",
       "      <td>9.179909</td>\n",
       "      <td>0.424138</td>\n",
       "      <td>0.379310</td>\n",
       "      <td>16.216178</td>\n",
       "      <td>0.320690</td>\n",
       "      <td>0.448276</td>\n",
       "      <td>0.231034</td>\n",
       "      <td>0.058621</td>\n",
       "      <td>0.022414</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.661111</td>\n",
       "      <td>0.088889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P49703</td>\n",
       "      <td>7.651052</td>\n",
       "      <td>0.417910</td>\n",
       "      <td>0.402985</td>\n",
       "      <td>14.652697</td>\n",
       "      <td>0.328358</td>\n",
       "      <td>0.477612</td>\n",
       "      <td>0.194030</td>\n",
       "      <td>0.099502</td>\n",
       "      <td>0.004975</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.567010</td>\n",
       "      <td>0.195876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20374</th>\n",
       "      <td>Q96H40</td>\n",
       "      <td>8.854868</td>\n",
       "      <td>0.330454</td>\n",
       "      <td>0.498920</td>\n",
       "      <td>15.918914</td>\n",
       "      <td>0.399568</td>\n",
       "      <td>0.494600</td>\n",
       "      <td>0.105832</td>\n",
       "      <td>0.032397</td>\n",
       "      <td>0.056156</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.659664</td>\n",
       "      <td>0.105042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20376</th>\n",
       "      <td>Q9BTX3</td>\n",
       "      <td>7.434628</td>\n",
       "      <td>0.514451</td>\n",
       "      <td>0.306358</td>\n",
       "      <td>14.472938</td>\n",
       "      <td>0.624277</td>\n",
       "      <td>0.352601</td>\n",
       "      <td>0.023121</td>\n",
       "      <td>0.104046</td>\n",
       "      <td>0.011561</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.460674</td>\n",
       "      <td>0.382022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20377</th>\n",
       "      <td>A6NFC5</td>\n",
       "      <td>7.800900</td>\n",
       "      <td>0.560538</td>\n",
       "      <td>0.179372</td>\n",
       "      <td>14.739747</td>\n",
       "      <td>0.573991</td>\n",
       "      <td>0.300448</td>\n",
       "      <td>0.125561</td>\n",
       "      <td>0.121076</td>\n",
       "      <td>0.031390</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.310000</td>\n",
       "      <td>0.420000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20378</th>\n",
       "      <td>Q8WV44</td>\n",
       "      <td>9.299208</td>\n",
       "      <td>0.365079</td>\n",
       "      <td>0.453968</td>\n",
       "      <td>16.340760</td>\n",
       "      <td>0.296825</td>\n",
       "      <td>0.528571</td>\n",
       "      <td>0.174603</td>\n",
       "      <td>0.077778</td>\n",
       "      <td>0.030159</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.658824</td>\n",
       "      <td>0.132353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20379</th>\n",
       "      <td>A0PK05</td>\n",
       "      <td>8.103288</td>\n",
       "      <td>0.458182</td>\n",
       "      <td>0.323636</td>\n",
       "      <td>15.087871</td>\n",
       "      <td>0.316364</td>\n",
       "      <td>0.683636</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.087273</td>\n",
       "      <td>0.021818</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.364078</td>\n",
       "      <td>0.412621</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16790 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id    length  hydr_count  polar_count  molecular_weight     helix  \\\n",
       "2      Q92667  9.818582    0.376523     0.370986         16.793404  0.079734   \n",
       "4      P62736  8.558421    0.427056     0.379310         15.574035  0.445623   \n",
       "5      Q9H553  8.700440    0.471154     0.358173         15.735720  0.485577   \n",
       "6      P0C7M7  9.179909    0.424138     0.379310         16.216178  0.320690   \n",
       "7      P49703  7.651052    0.417910     0.402985         14.652697  0.328358   \n",
       "...       ...       ...         ...          ...               ...       ...   \n",
       "20374  Q96H40  8.854868    0.330454     0.498920         15.918914  0.399568   \n",
       "20376  Q9BTX3  7.434628    0.514451     0.306358         14.472938  0.624277   \n",
       "20377  A6NFC5  7.800900    0.560538     0.179372         14.739747  0.573991   \n",
       "20378  Q8WV44  9.299208    0.365079     0.453968         16.340760  0.296825   \n",
       "20379  A0PK05  8.103288    0.458182     0.323636         15.087871  0.316364   \n",
       "\n",
       "           turn     sheet         A         C  ...  Methylation_MSD  \\\n",
       "2      0.805094  0.115172  0.079734  0.018826  ...              0.0   \n",
       "4      0.347480  0.206897  0.076923  0.018568  ...              1.0   \n",
       "5      0.375000  0.139423  0.072115  0.028846  ...              1.0   \n",
       "6      0.448276  0.231034  0.058621  0.022414  ...              0.0   \n",
       "7      0.477612  0.194030  0.099502  0.004975  ...              1.0   \n",
       "...         ...       ...       ...       ...  ...              ...   \n",
       "20374  0.494600  0.105832  0.032397  0.056156  ...              0.0   \n",
       "20376  0.352601  0.023121  0.104046  0.011561  ...              1.0   \n",
       "20377  0.300448  0.125561  0.121076  0.031390  ...              1.0   \n",
       "20378  0.528571  0.174603  0.077778  0.030159  ...              1.0   \n",
       "20379  0.683636  0.000000  0.087273  0.021818  ...              0.0   \n",
       "\n",
       "       coiled_coil  EGF  RAS_profile  RRM  ww_domain  transmembrane  TMHMM  \\\n",
       "2                1    0            0    0          0              1      1   \n",
       "4                0    0            0    0          0              0      0   \n",
       "5                0    0            0    0          0              1      0   \n",
       "6                0    0            0    0          0              0      0   \n",
       "7                0    0            0    0          0              0      0   \n",
       "...            ...  ...          ...  ...        ...            ...    ...   \n",
       "20374            0    0            0    0          0              0      0   \n",
       "20376            0    0            0    0          0              1      1   \n",
       "20377            0    0            0    0          0              1      1   \n",
       "20378            0    0            0    0          0              0      0   \n",
       "20379            0    0            0    0          0              1      1   \n",
       "\n",
       "       Polar_exposed  Hydrophobic_exposed  \n",
       "2           0.447738             0.263651  \n",
       "4           0.682540             0.111111  \n",
       "5           0.614286             0.185714  \n",
       "6           0.661111             0.088889  \n",
       "7           0.567010             0.195876  \n",
       "...              ...                  ...  \n",
       "20374       0.659664             0.105042  \n",
       "20376       0.460674             0.382022  \n",
       "20377       0.310000             0.420000  \n",
       "20378       0.658824             0.132353  \n",
       "20379       0.364078             0.412621  \n",
       "\n",
       "[16790 rows x 96 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_MS_filtered # filtered human proteome dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>length</th>\n",
       "      <th>hydr_count</th>\n",
       "      <th>polar_count</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>helix</th>\n",
       "      <th>turn</th>\n",
       "      <th>sheet</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>...</th>\n",
       "      <th>Methylation_MSD</th>\n",
       "      <th>coiled_coil</th>\n",
       "      <th>EGF</th>\n",
       "      <th>RAS_profile</th>\n",
       "      <th>RRM</th>\n",
       "      <th>ww_domain</th>\n",
       "      <th>transmembrane</th>\n",
       "      <th>TMHMM</th>\n",
       "      <th>Polar_exposed</th>\n",
       "      <th>Hydrophobic_exposed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q8N7X0</td>\n",
       "      <td>10.703038</td>\n",
       "      <td>0.376125</td>\n",
       "      <td>0.431314</td>\n",
       "      <td>17.745289</td>\n",
       "      <td>0.225555</td>\n",
       "      <td>0.604079</td>\n",
       "      <td>0.170366</td>\n",
       "      <td>0.056989</td>\n",
       "      <td>0.011398</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.597984</td>\n",
       "      <td>0.161254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q5T1N1</td>\n",
       "      <td>9.707359</td>\n",
       "      <td>0.290670</td>\n",
       "      <td>0.479665</td>\n",
       "      <td>16.719373</td>\n",
       "      <td>0.183014</td>\n",
       "      <td>0.777512</td>\n",
       "      <td>0.039474</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.021531</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.570968</td>\n",
       "      <td>0.162903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q92667</td>\n",
       "      <td>9.818582</td>\n",
       "      <td>0.376523</td>\n",
       "      <td>0.370986</td>\n",
       "      <td>16.793404</td>\n",
       "      <td>0.079734</td>\n",
       "      <td>0.805094</td>\n",
       "      <td>0.115172</td>\n",
       "      <td>0.079734</td>\n",
       "      <td>0.018826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.447738</td>\n",
       "      <td>0.263651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q5VUY0</td>\n",
       "      <td>8.668885</td>\n",
       "      <td>0.511057</td>\n",
       "      <td>0.312039</td>\n",
       "      <td>15.706400</td>\n",
       "      <td>0.491400</td>\n",
       "      <td>0.380835</td>\n",
       "      <td>0.127764</td>\n",
       "      <td>0.058968</td>\n",
       "      <td>0.039312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.560000</td>\n",
       "      <td>0.256000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P62736</td>\n",
       "      <td>8.558421</td>\n",
       "      <td>0.427056</td>\n",
       "      <td>0.379310</td>\n",
       "      <td>15.574035</td>\n",
       "      <td>0.445623</td>\n",
       "      <td>0.347480</td>\n",
       "      <td>0.206897</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.018568</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.682540</td>\n",
       "      <td>0.111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20379</th>\n",
       "      <td>A0PK05</td>\n",
       "      <td>8.103288</td>\n",
       "      <td>0.458182</td>\n",
       "      <td>0.323636</td>\n",
       "      <td>15.087871</td>\n",
       "      <td>0.316364</td>\n",
       "      <td>0.683636</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.087273</td>\n",
       "      <td>0.021818</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.364078</td>\n",
       "      <td>0.412621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20380</th>\n",
       "      <td>Q9HCN2</td>\n",
       "      <td>6.954196</td>\n",
       "      <td>0.370968</td>\n",
       "      <td>0.282258</td>\n",
       "      <td>13.887011</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.088710</td>\n",
       "      <td>0.016129</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.397590</td>\n",
       "      <td>0.156627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20381</th>\n",
       "      <td>A0A0A0MS03</td>\n",
       "      <td>6.832890</td>\n",
       "      <td>0.412281</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>13.836785</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.535088</td>\n",
       "      <td>0.438596</td>\n",
       "      <td>0.061404</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.388060</td>\n",
       "      <td>0.268657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20382</th>\n",
       "      <td>A0A0A6YYK4</td>\n",
       "      <td>6.845490</td>\n",
       "      <td>0.417391</td>\n",
       "      <td>0.347826</td>\n",
       "      <td>13.838505</td>\n",
       "      <td>0.026087</td>\n",
       "      <td>0.573913</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.104348</td>\n",
       "      <td>0.026087</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.449275</td>\n",
       "      <td>0.275362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20383</th>\n",
       "      <td>A0A0A0MS05</td>\n",
       "      <td>6.832890</td>\n",
       "      <td>0.394737</td>\n",
       "      <td>0.315789</td>\n",
       "      <td>13.815908</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.543860</td>\n",
       "      <td>0.429825</td>\n",
       "      <td>0.043860</td>\n",
       "      <td>0.035088</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>0.261538</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20381 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               id     length  hydr_count  polar_count  molecular_weight  \\\n",
       "0          Q8N7X0  10.703038    0.376125     0.431314         17.745289   \n",
       "1          Q5T1N1   9.707359    0.290670     0.479665         16.719373   \n",
       "2          Q92667   9.818582    0.376523     0.370986         16.793404   \n",
       "3          Q5VUY0   8.668885    0.511057     0.312039         15.706400   \n",
       "4          P62736   8.558421    0.427056     0.379310         15.574035   \n",
       "...           ...        ...         ...          ...               ...   \n",
       "20379      A0PK05   8.103288    0.458182     0.323636         15.087871   \n",
       "20380      Q9HCN2   6.954196    0.370968     0.282258         13.887011   \n",
       "20381  A0A0A0MS03   6.832890    0.412281     0.315789         13.836785   \n",
       "20382  A0A0A6YYK4   6.845490    0.417391     0.347826         13.838505   \n",
       "20383  A0A0A0MS05   6.832890    0.394737     0.315789         13.815908   \n",
       "\n",
       "          helix      turn     sheet         A         C  ...  Methylation_MSD  \\\n",
       "0      0.225555  0.604079  0.170366  0.056989  0.011398  ...              1.0   \n",
       "1      0.183014  0.777512  0.039474  0.052632  0.021531  ...              1.0   \n",
       "2      0.079734  0.805094  0.115172  0.079734  0.018826  ...              0.0   \n",
       "3      0.491400  0.380835  0.127764  0.058968  0.039312  ...              0.0   \n",
       "4      0.445623  0.347480  0.206897  0.076923  0.018568  ...              1.0   \n",
       "...         ...       ...       ...       ...       ...  ...              ...   \n",
       "20379  0.316364  0.683636  0.000000  0.087273  0.021818  ...              0.0   \n",
       "20380  0.000000  0.750000  0.250000  0.088710  0.016129  ...              1.0   \n",
       "20381  0.026316  0.535088  0.438596  0.061404  0.035088  ...              0.0   \n",
       "20382  0.026087  0.573913  0.400000  0.104348  0.026087  ...              0.0   \n",
       "20383  0.026316  0.543860  0.429825  0.043860  0.035088  ...              0.0   \n",
       "\n",
       "       coiled_coil  EGF  RAS_profile  RRM  ww_domain  transmembrane  TMHMM  \\\n",
       "0                0    0            0    0          0              0      0   \n",
       "1                0    0            0    0          0              0      0   \n",
       "2                1    0            0    0          0              1      1   \n",
       "3                0    0            0    0          0              0      1   \n",
       "4                0    0            0    0          0              0      0   \n",
       "...            ...  ...          ...  ...        ...            ...    ...   \n",
       "20379            0    0            0    0          0              1      1   \n",
       "20380            0    0            0    0          0              0      0   \n",
       "20381            0    0            0    0          0              1      0   \n",
       "20382            0    0            0    0          0              1      0   \n",
       "20383            0    0            0    0          0              1      0   \n",
       "\n",
       "       Polar_exposed  Hydrophobic_exposed  \n",
       "0           0.597984             0.161254  \n",
       "1           0.570968             0.162903  \n",
       "2           0.447738             0.263651  \n",
       "3           0.560000             0.256000  \n",
       "4           0.682540             0.111111  \n",
       "...              ...                  ...  \n",
       "20379       0.364078             0.412621  \n",
       "20380       0.397590             0.156627  \n",
       "20381       0.388060             0.268657  \n",
       "20382       0.449275             0.275362  \n",
       "20383       0.400000             0.261538  \n",
       "\n",
       "[20381 rows x 96 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_log # unfiltered human proteome dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save final unfiltered feature dataset\n",
    "df_log.to_csv(Data_path + '/Curated/features_human_proteome_no_filtering.csv', index=False)\n",
    "# save final MS filtered feature dataset\n",
    "df_MS_filtered.to_csv(Data_path + '/Curated/features_human_proteome.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
